pkg/proxy/winuserspace/proxier.go

```diff
@@ -304,322 +319,79 @@ func (proxier *Proxier) OnServiceUpdate(services []api.Service) {
-	}
-	if len(el) == 0 {
-		glog.V(3).Infof("Closed iptables portals for service %q", service)
-	} else {
-		glog.Errorf("Some errors closing iptables portals for service %q", service)
+	// servicePortNameMap tracks all service port portals with the same name/port.
+	// A value of true means there is one or more service port portals with name/port pair.
+	servicePortNameMap := make(map[proxy.ServicePortName]bool)
+	for name := range proxier.serviceMap {
+		servicePortName := proxy.ServicePortName{NamespacedName: types.NamespacedName{Namespace: name.Namespace, Name: name.Name}, Port: name.Port}
+		servicePortNameMap[servicePortName] = servicePortNameMap[servicePortName] || activeServicePortPortals[name]
 	}
-	return utilerrors.NewAggregate(el)
-}
-
-func (proxier *Proxier) closeOnePortal(portal portal, protocol api.Protocol, proxyIP net.IP, proxyPort int, name proxy.ServicePortName) []error {
-	el := []error{}
 
-	if local, err := isLocalIP(portal.ip); err != nil {
-		el = append(el, fmt.Errorf("can't determine if IP is local, assuming not: %v", err))
-	} else if local {
-		if err := proxier.releaseNodePort(portal.ip, portal.port, protocol, name); err != nil {
-			el = append(el, err)
+	// Only delete load balancer if all listen ips per name/port show inactive.
+	for name := range servicePortNameMap {
+		if !servicePortNameMap[name] {
+			proxier.loadBalancer.DeleteService(name)
 		}
 	}
-
-	args := proxier.netshIpv4AddressDeleteArgs(portal.ip)
-	if err := proxier.netsh.DeleteIPAddress(args); err != nil {
-		glog.Errorf("Failed to delete IP address for service %q", name)
-		el = append(el, err)
-	}
-
-	args = proxier.netshPortProxyDeleteArgs(portal.ip, portal.port, proxyIP, proxyPort, name)
-	if err := proxier.netsh.DeletePortProxyRule(args); err != nil {
-		glog.Errorf("Failed to delete portproxy rule for service %q", name)
-		el = append(el, err)
-	}
-
-	return el
-}
-
-func (proxier *Proxier) closeNodePort(nodePort int, protocol api.Protocol, proxyIP net.IP, proxyPort int, name proxy.ServicePortName) []error {
-	el := []error{}
-
-	args := proxier.netshPortProxyDeleteArgs(nil, nodePort, proxyIP, proxyPort, name)
-	if err := proxier.netsh.DeletePortProxyRule(args); err != nil {
-		glog.Errorf("Failed to delete portproxy rule for service %q", name)
-		el = append(el, err)
-	}
-
-	if err := proxier.releaseNodePort(nil, nodePort, protocol, name); err != nil {
-		el = append(el, err)
-	}
-
-	return el
 }
 
-func isLocalIP(ip net.IP) (bool, error) {
-	addrs, err := net.InterfaceAddrs()
-	if err != nil {
-		return false, err
+func sameConfig(info *serviceInfo, service *api.Service, protocol api.Protocol, listenPort int) bool {
+	if info.protocol != protocol || info.portal.port != listenPort {
+		return false
 	}
-	for i := range addrs {
-		intf, _, err := net.ParseCIDR(addrs[i].String())
-		if err != nil {
-			return false, err
-		}
-		if ip.Equal(intf) {
-			return true, nil
-		}
+	if info.sessionAffinityType != service.Spec.SessionAffinity {
```

Comment:
It feels kind of wierd to have this be a separate condition.

Either three different `if` or one large `if` but in between feels weird.